src/modelfree/transparent.py

```diff
@@ -0,0 +1,188 @@
+                 hiddens=None, scope="input", reuse=False, normalize=False):
+        MlpPolicyValue.__init__(self, sess, ob_space, ac_space, n_env, n_steps, n_batch,
+                                hiddens=hiddens, scope=scope, reuse=reuse, normalize=normalize)
+        TransparentPolicy.__init__(self, transparent_params)
+
+    def get_obs_sizes(self):
+        return self.ob_space.shape[0], sum(self.hiddens), sum(self.hiddens), None
+
+    def step(self, obs, state=None, mask=None, deterministic=False):
+        action = self.deterministic_action if deterministic else self.action
+        outputs = [action, self.value_flat, self.neglogp, self.ff_out]
+        a, v, neglogp, ff = self.sess.run(outputs, {self.obs_ph: obs})
+
+        transparent_objs = (obs, np.concatenate(ff['policy']), np.concatenate(ff['value']), None)
+        transparency_dict = self.get_transparency_dict(transparent_objs)
+        return a, v, self.initial_state, neglogp, transparency_dict
+
+
+class TransparentCurryVecEnv(CurryVecEnv):
+    """CurryVecEnv that provides transparency data about its policy in its infos dicts."""
+    def __init__(self, venv, policy, agent_idx=0):
+        super().__init__(venv, policy, agent_idx)
+        self.underlying_policy = policy.policy
+        if not isinstance(self.underlying_policy, TransparentPolicy):
+            raise TypeError("Error: policy must be transparent")
+        self._action = None
+
+        obs_aug_amount = self.underlying_policy.get_obs_aug_amount()
+        if obs_aug_amount > 0:
+            obs_aug_space = gym.spaces.Box(-np.inf, np.inf, (obs_aug_amount,))
+            self.observation_space = _tuple_space_augment(self.observation_space, agent_idx,
+                                                          augment_space=obs_aug_space)
+
+    def step_async(self, actions):
+        actions.insert(self._agent_to_fix, self._action)
+        self.venv.step_async(actions)
+
+    def step_wait(self):
+        observations, rewards, self._dones, infos = self.venv.step_wait()
+        observations = self._get_updated_obs(observations)
+        for env_idx in range(self.num_envs):
+            env_data = {k: v[env_idx] for k, v in self._data.items()}
+            infos[env_idx][self._agent_to_fix].update(env_data)
+        return observations, rewards, self._dones, infos
+
+    def reset(self):
+        observations = self._get_updated_obs(self.venv.reset())
+        return observations
+
+    def _get_updated_obs(self, observations):
+        observations, self._obs = _tuple_pop(observations, self._agent_to_fix)
+        self._action, self._state, self._data = self._policy.predict(self._obs, state=self._state,
+                                                                     mask=self._dones)
+        # we assume that there is only one other agent in the MultiEnv.
+        assert len(observations) == 1
+        obs_copy = observations[0]
+        for k in TRANSPARENCY_KEYS:
```

Comment:
Would suggest iterating over `self._data.items()` instead to reduce checks needed in the next line.